
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">



<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-19801227-1']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
    <title>evol fetch &mdash; RuleBender</title>
    <link rel="stylesheet" href="../../../../_static/css/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/css/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/css/bootswatch-united.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/css/rbdocs.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../../',
        VERSION:     '1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  false
      };
    </script>
    <script type="text/javascript" src="../../../../_static/js/jquery-2.0.0.min.js"></script>
    <script type="text/javascript" src="../../../../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../../_static/js/zeroclipboard.min.js"></script>
    <script type="text/javascript" src="../../../../_static/js/copybutton.js"></script>
    <script type="text/javascript" src="../../../../_static/js/rbdocs.js"></script>
    <link rel="shortcut icon" href="../../../../_static/bionetgen.ico"/>

  </head>
  <body data-spy="scroll" data-target="#localtoc" data-offset="190">

<div class="container">
  <div class="row">
    <div class="span12">
      <a href="../../../../index.html">
        <img src="../../../../_static/bionetgen_logo.png" alt="BNG logo" id="logo" />
      </a>
    </div>
    <div class="span12">
      <div class="subnav">
        <ul class="nav nav-pills">
        <li class="pull-right">
          <a href="../../../../genindex.html" title="General Index"
             accesskey="I">index</a>
          </li>
        <li><a href="../../../../index.html">Home  &raquo;</a></li> 
        </ul>
      </div>
    </div>
  </div>

  <div class="content">
  <div class="row">
    <div class="span3">
      <div id="sidebar" style="width: 220px">
  <div id="localtoc" style="margin-bottom: 20px">
    <ul>
<li><a class="reference internal" href="#">evol fetch</a><ul>
<li><a class="reference internal" href="#usage">Usage</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
</ul>

  </div>
        <div class="accordion" id="accordion2">
  <div id="searchbox" class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseToolbox">
        <strong>Toolbox</strong>
      </a>
    </div>
    <div id="collapseToolbox" class="accordion-body collapse in">
      <div class="accordion-inner">
        <div id="showcodebuttons" style="margin-bottom: 10px">
          <p><small>Show code snippets for copying:</small></p>
          <div class="btn-group">
            <a href="#codemodal" role="button" data-toggle="modal" id="showcode" class="btn btn-primary btn-small">
              <i class="icon-fullscreen icon-white"></i> Show code</a>
            <a href="#codemodal" role="button" data-toggle="modal" id="withcomments" class="btn btn-primary btn-small">
              and output <i class="icon-comment icon-white"></i> </a>
          </div>
        </div>

        <p><small>Search terms or function by name:</small></p>
        <form class="form-search pull-right" action="../../../../search.html" method="get">
          <div class="input-append">
            <input type="text" class="input-small search-query" name="q" class="span2">
            <button type="submit" class="btn btn-primary">Go</button>
          </div>
          <input type="hidden" name="check_keywords" value="yes" />
          <input type="hidden" name="area" value="default" />
        </form>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseReleaseNotes">
        <strong>Release Notes</strong>
      </a>
    </div>
    <div id="collapseReleaseNotes" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small><strong>v1</strong> series come with new and improved sequence, structure, and dynamics analysis
features. See <a href="../../../../manual/release/index.html">release notes</a> for details.
        </small></p>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseHowToCite">
        <strong>How to Cite</strong>
      </a>
    </div>
    <div id="collapseHowToCite" class="accordion-body collapse in">
      <div class="accordion-inner">
                      <p>
                <small>
                  <strong>BioNetGen</strong> <br>
                  J. R. Faeder, M. L. Blinov, and W. S. Hlavacek (2009) <a href="http://www.ncbi.nlm.nih.gov/pubmed/19399430/">Rule-based modeling of biochemical systems with BioNetGen</a>, <i>Methods Mol. Biol.</i>, <strong>500</strong>, 113-67. <br> 
                 </small>
              </p>  
              <p>
                <small>
                 <strong>RuleBender</strong> <br>
                  A. M. Smith, W. Xu, Y. Sun, J. R. Faeder and G. E. Marai (2012) <a href="http://bmcbioinformatics.biomedcentral.com:80/articles/10.1186/1471-2105-13-S8-S3">RuleBender: integrated modeling, simulation and visualization for rule-based intracellular biochemistry</a>, <i>BMC Bioinformatics</i>, <strong>13</strong>(suppl. 8), S3 (16 pages). <br>
                </small>
              </p>  
              <p>
                <small>
                  <strong>NFsim</strong> <br>
                  M. W. Sneddon, J. R. Faeder and T. Emonet (2012) <a href="https://www.ncbi.nlm.nih.gov/pubmed/21186362">Efficient modeling, simulation and coarse-graining of biological complexity with NFsim.</a>, <i>Nat. Methods</i>, <strong>8</strong>, 177-183. 
                </small>
              </p>
      </div>
    </div>
  </div>
        </div>
      </div>
    </div>
    <div class="span9 body">
    
  <div class="section" id="evol-fetch">
<span id="id1"></span><h1>evol fetch<a class="headerlink" href="#evol-fetch" title="Permalink to this headline">¶</a></h1>
<div class="section" id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p>Running <strong class="command">evol fetch -h</strong> displays:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>usage: evol fetch [-h] [--quiet] [--examples] [-a STR] [-f STR] [-o STR]
                  [-i STR] [-g STR] [-t INT] [-d PATH] [-p STR] [-z]
                  acc

positional arguments:
  acc                   Pfam accession or ID

optional arguments:
  -h, --help            show this help message and exit
  --quiet               suppress info messages to stderr
  --examples            show usage examples and exit

download options:
  -a STR, --alignment STR
                        alignment type, one of full, seed, ncbi, metagenomics
                        (default: full)
  -f STR, --format STR  Pfam supported MSA format, one of selex, fasta,
                        stockholm (default: selex)
  -o STR, --order STR   ordering of sequences, one of tree, alphabetical
                        (default: tree)
  -i STR, --inserts STR
                        letter case for inserts, one of upper, lower (default:
                        upper)
  -g STR, --gaps STR    gap character, one of dashes, dots, mixed (default:
                        dashes)
  -t INT, --timeout INT
                        timeout for blocking connection attempts (default: 60)

output options:
  -d PATH, --outdir PATH
                        output directory (default: .)
  -p STR, --outname STR
                        output filename, default is accession and alignment
                        type
  -z, --compressed      gzip downloaded MSA file
</pre></div>
</div>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<p>Running <strong class="command">evol fetch --examples</strong> displays:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>Sequence coevolution analysis involves several steps that including
retrieving data and refining it for calculations.  These steps are
illustrated below for RnaseA protein family.

Search Pfam database:

  $  evol search 2w5i

Download Pfam MSA file:

  $  evol fetch RnaseA

Refine MSA file:

  $ evol refine RnaseA_full.slx -l RNAS1_BOVIN --seqid 0.98 --rowocc 0.8

Checking occupancy:

  $ evol occupancy RnaseA_full.slx -l RNAS1_BOVIN -o col -S

Conservation analysis:

  $ evol conserv RnaseA_full_refined.slx

Coevolution analysis:

  $ evol coevol RnaseA_full_refined.slx -S -c apc

Rank order analysis:

  $ evol rankorder RnaseA_full_refined_mutinfo_corr_apc.txt -p 2w5i_1-121.pdb --seq-sep 3
</pre></div>
</div>
</div>
</div>


    </div>
  </div>
</div>
<footer class="footer">
  <div class="row">
    <div class="span12">
      <p><small>
      &copy; Copyright 2010-2017, University of Pittsburgh.
      Last updated on Dec 07, 2017.
      Created using <a href="http://sphinx-doc.org">Sphinx</a> 1.3.5.
      </small></p>
    </div>
  </div>
  </div>
</footer>
<div id="codemodal" class="modal hide fade" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
  <div class="modal-header">
    <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
    <h3 id="myModalLabel">Code Snippets</h3>
  </div>
  <div class="modal-body">
    <textarea id="codesnippets"></textarea>
  </div>
  <div class="modal-footer">
    <button class="btn btn-info" id="selectcode">Select Code</button>
    <!--
    <button class="btn btn-success" id="copycode" data-clipboard-target="codesnippets">Copy Code</button>
    -->
    <button class="btn" data-dismiss="modal" aria-hidden="true">Close</button>
  </div>
  <input type="hidden" id="zeroclipboardpath" value="../../../../_static/js/zeroclipboard.swf">
</div>
  </body>
</html>