
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">



<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-19801227-1']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
    <title>Sequence Analysis &mdash; RuleBender</title>
    <link rel="stylesheet" href="../../_static/css/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/css/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/css/bootswatch-united.min.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/css/prodydocs.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  false
      };
    </script>
    <script type="text/javascript" src="../../_static/js/jquery-2.0.0.min.js"></script>
    <script type="text/javascript" src="../../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../_static/js/zeroclipboard.min.js"></script>
    <script type="text/javascript" src="../../_static/js/copybutton.js"></script>
    <script type="text/javascript" src="../../_static/js/prodydocs.js"></script>
    <link rel="shortcut icon" href="../../_static/favicon.ico"/>

  </head>
  <body data-spy="scroll" data-target="#localtoc" data-offset="190">

<div class="container">
  <div class="row">
    <div class="span12">
      <a href="../../index.html">
        <img src="../../_static/logo.png" alt="ProDy logo" id="logo" />
      </a>
    </div>
    <div class="span12">
      <div class="subnav">
        <ul class="nav nav-pills">
        <li class="pull-right">
          <a href="../../genindex.html" title="General Index"
             accesskey="I">index</a>
          </li>
        <li class="pull-right">
          <a href="../../py-modindex.html" title="Python Module Index"
             >modules</a>
          </li>
        <li><a href="../../index.html">Home  &raquo;</a></li> 
        </ul>
      </div>
    </div>
  </div>

  <div class="content">
  <div class="row">
    <div class="span3">
      <div id="sidebar" style="width: 220px">
  <div id="localtoc" style="margin-bottom: 20px">
    <ul>
<li><a class="reference internal" href="#">Sequence Analysis</a><ul>
<li><a class="reference internal" href="#access-pfam">Access Pfam</a></li>
<li><a class="reference internal" href="#parse-msa">Parse MSA</a></li>
<li><a class="reference internal" href="#sequences">Sequences</a></li>
<li><a class="reference internal" href="#analysis">Analysis</a></li>
</ul>
</li>
</ul>

  </div>
        <div class="accordion" id="accordion2">
  <div id="searchbox" class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseToolbox">
        <strong>Toolbox</strong>
      </a>
    </div>
<!--     <div id="collapseToolbox" class="accordion-body collapse in">
      <div class="accordion-inner">
        <div id="showcodebuttons" style="margin-bottom: 10px">
          <p><small>Show code snippets for copying:</small></p>
          <div class="btn-group">
            <a href="#codemodal" role="button" data-toggle="modal" id="showcode" class="btn btn-primary btn-small">
              <i class="icon-fullscreen icon-white"></i> Show code</a>
            <a href="#codemodal" role="button" data-toggle="modal" id="withcomments" class="btn btn-primary btn-small">
              and output <i class="icon-comment icon-white"></i> </a>
          </div>
        </div> -->

        <p><small>Search terms or function by name:</small></p>
        <form class="form-search pull-right" action="../../search.html" method="get">
          <div class="input-append">
            <input type="text" class="input-small search-query" name="q" class="span2">
            <button type="submit" class="btn btn-primary">Go</button>
          </div>
          <input type="hidden" name="check_keywords" value="yes" />
          <input type="hidden" name="area" value="default" />
        </form>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseReleaseNotes">
        <strong>Release Notes</strong>
      </a>
    </div>
    <div id="collapseReleaseNotes" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small><strong>v1</strong> series come with new and improved sequence, structure, and dynamics analysis
features. See <a href="../../manual/release/index.html">release notes</a> for details.
        </small></p>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseHowToCite">
        <strong>How to Cite</strong>
      </a>
    </div>
    <div id="collapseHowToCite" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small>
          Bakan A, Meireles LM, Bahar I
          <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/27/11/1575" target="_blank">
          <i>ProDy</i>: Protein Dynamics Inferred from Theory and Experiments</a> <br/>
          <em>Bioinformatics</em> <strong>2011</strong> 27(11):1575-1577.
        </small></p>
      </div>
    </div>
  </div>
        </div>
      </div>
    </div>
    <div class="span9 body">
    
  <div class="section" id="sequence-analysis">
<h1>Sequence Analysis<a class="headerlink" href="#sequence-analysis" title="Permalink to this headline">¶</a></h1>
<p>Evol component of ProDy package brought new fast, flexible, and efficient
features for handling multiple sequence alignments and analyzing sequence
evolution.  Here, we just give a brief introduction to these features.  For
more detailed examples, see <span class="xref std std-ref">evol-tutorial</span>.</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">from</span> <span class="nn">prody</span> <span class="kn">import</span> <span class="o">*</span>

<span class="gp">In [2]: </span><span class="kn">from</span> <span class="nn">pylab</span> <span class="kn">import</span> <span class="o">*</span>

<span class="gp">In [3]: </span><span class="n">ion</span><span class="p">()</span>
</pre></div>
</div>
<div class="section" id="access-pfam">
<h2>Access Pfam<a class="headerlink" href="#access-pfam" title="Permalink to this headline">¶</a></h2>
<p>First, let&#8217;s fetch an MSA file from <a class="reference external" href="http://pfam.sanger.ac.uk/">Pfam</a> database:</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [4]: </span><span class="n">filename</span> <span class="o">=</span> <span class="n">fetchPfamMSA</span><span class="p">(</span><span class="s1">&#39;pkinase&#39;</span><span class="p">,</span> <span class="n">alignment</span><span class="o">=</span><span class="s1">&#39;seed&#39;</span><span class="p">)</span>

<span class="gp">In [5]: </span><span class="n">filename</span>
<span class="gh">Out[5]: </span><span class="go">&#39;pkinase_seed.sth&#39;</span>
</pre></div>
</div>
<p>We downloaded the seed alignment for Protein Kinase (<a class="reference external" href="http://pfam.sanger.ac.uk/family/Pkinase">Pkinase</a>) family.</p>
</div>
<div class="section" id="parse-msa">
<h2>Parse MSA<a class="headerlink" href="#parse-msa" title="Permalink to this headline">¶</a></h2>
<p>As you might guess, we will parse this file using <a class="reference internal" href="../../manual/reference/sequence/msafile.html#prody.sequence.msafile.parseMSA" title="prody.sequence.msafile.parseMSA"><code class="xref py py-func docutils literal"><span class="pre">parseMSA()</span></code></a> function:</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [6]: </span><span class="n">msa</span> <span class="o">=</span> <span class="n">parseMSA</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

<span class="gp">In [7]: </span><span class="n">msa</span>
<span class="gh">Out[7]: </span><span class="go">&lt;MSA: pkinase_seed (38 sequences, 419 residues)&gt;</span>
</pre></div>
</div>
</div>
<div class="section" id="sequences">
<h2>Sequences<a class="headerlink" href="#sequences" title="Permalink to this headline">¶</a></h2>
<p>You can access <a class="reference internal" href="../../manual/reference/sequence/sequence.html#prody.sequence.sequence.Sequence" title="prody.sequence.sequence.Sequence"><code class="xref py py-class docutils literal"><span class="pre">Sequence</span></code></a> objects by indexing <a class="reference internal" href="../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a>:</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [8]: </span><span class="n">seq</span> <span class="o">=</span> <span class="n">msa</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="gp">In [9]: </span><span class="n">seq</span>
<span class="gh">Out[9]: </span><span class="go">&lt;Sequence: CDC15_YEAST (pkinase_seed[0]; length 419; 248 residues and 171 gaps)&gt;</span>

<span class="gp">In [10]: </span><span class="k">print</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="go">YHLKQVIGRGSYGVV..YKAINKHTDQVVAIKEVVYENDEELND........IMAEISLLKNLN...HNNIVKYHGFIRK.............SYELY..ILLEYCANGS....LRRLIS.......RSSTGLSEN.......ESK.TYVTQTLLGLKYLH.GEG...VIHRDIKAANILLS............ADN.......TVKLADFGVSTIVNSS.....................ALTLAGTLNWMAPEIL..........GN...RGA.STLSDIWSLGATVVEMLTKNPPYHNLTDANIYYAVEND.........................TYYPPSSF....................................SEPLKDFLSKCFVKNMYKRPT.....ADQLLKHVWI</span>
</pre></div>
</div>
<p>You can also slice <a class="reference internal" href="../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> objects and iterate over sequences:</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [11]: </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">msa</span><span class="p">[:</span><span class="mi">4</span><span class="p">]:</span>
<span class="gp">   ....: </span>    <span class="nb">repr</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
<span class="gp">   ....: </span>
</pre></div>
</div>
</div>
<div class="section" id="analysis">
<h2>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">¶</a></h2>
<p>Evol component includes several functions for calculating conservation and
coevolution properties of amino acids, which are shown in <span class="xref std std-ref">evol-tutorial</span>.
Here, let&#8217;s take a look at <a class="reference internal" href="../../manual/reference/sequence/analysis.html#prody.sequence.analysis.calcMSAOccupancy" title="prody.sequence.analysis.calcMSAOccupancy"><code class="xref py py-func docutils literal"><span class="pre">calcMSAOccupancy()</span></code></a> and
<a class="reference internal" href="../../manual/reference/sequence/plotting.html#prody.sequence.plotting.showMSAOccupancy" title="prody.sequence.plotting.showMSAOccupancy"><code class="xref py py-func docutils literal"><span class="pre">showMSAOccupancy()</span></code></a> functions:</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [12]: </span><span class="n">occ</span> <span class="o">=</span> <span class="n">calcMSAOccupancy</span><span class="p">(</span><span class="n">msa</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="gp">In [13]: </span><span class="n">occ</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="gh">Out[13]: </span><span class="go">1.0</span>
</pre></div>
</div>
<p>This shows that an amino acid is present only in one of the sequences in the
MSA.</p>
<div class="highlight-ipython"><div class="highlight"><pre><span></span><span class="gp">In [14]: </span><span class="n">showMSAOccupancy</span><span class="p">(</span><span class="n">msa</span><span class="p">,</span> <span class="n">count</span><span class="o">=</span><span class="bp">True</span><span class="p">);</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/prody_tutorial_sequence_occ.png"><img alt="../../_images/prody_tutorial_sequence_occ.png" src="../../_images/prody_tutorial_sequence_occ.png" style="width: 6in;" /></a>
<p>You see that many residues are not present in all sequences. You will see
how to refine such MSA instances in <span class="xref std std-ref">evol-tutorial</span>.</p>
</div>
</div>


    </div>
  </div>
</div>
<footer class="footer">
  <div class="row">
    <div class="span12">
      <p><small>
      &copy; Copyright 2010-2017, University of Pittsburgh.
      Last updated on Dec 06, 2017.
      Created using <a href="http://sphinx-doc.org">Sphinx</a> 1.3.5.
      </small></p>
    </div>
  </div>
  </div>
</footer>
<div id="codemodal" class="modal hide fade" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
  <div class="modal-header">
    <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
    <h3 id="myModalLabel">Code Snippets</h3>
  </div>
  <div class="modal-body">
    <textarea id="codesnippets"></textarea>
  </div>
  <div class="modal-footer">
    <button class="btn btn-info" id="selectcode">Select Code</button>
    <!--
    <button class="btn btn-success" id="copycode" data-clipboard-target="codesnippets">Copy Code</button>
    -->
    <button class="btn" data-dismiss="modal" aria-hidden="true">Close</button>
  </div>
  <input type="hidden" id="zeroclipboardpath" value="../../_static/js/zeroclipboard.swf">
</div>
  </body>
</html>