
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">



<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-19801227-1']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
    <title>Analysis Functions &mdash; RuleBender</title>
    <link rel="stylesheet" href="../../../_static/css/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/css/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/css/bootswatch-united.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/css/prodydocs.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  false
      };
    </script>
    <script type="text/javascript" src="../../../_static/js/jquery-2.0.0.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/zeroclipboard.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/copybutton.js"></script>
    <script type="text/javascript" src="../../../_static/js/prodydocs.js"></script>
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>

  </head>
  <body data-spy="scroll" data-target="#localtoc" data-offset="190">

<div class="container">
  <div class="row">
    <div class="span12">
      <a href="../../../index.html">
        <img src="../../../_static/logo.png" alt="ProDy logo" id="logo" />
      </a>
    </div>
    <div class="span12">
      <div class="subnav">
        <ul class="nav nav-pills">
        <li class="pull-right">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a>
          </li>
        <li class="pull-right">
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a>
          </li>
        <li class="pull-right">
          <a href="msa.html" title="Multiple Sequence Alignment"
             accesskey="N">next</a>
          </li>
        <li class="pull-right">
          <a href="index.html" title="Sequence Analysis"
             accesskey="P">previous</a>
          </li>
        <li><a href="../../../index.html">Home  &raquo;</a></li>
          <li><a href="../../index.html" >ProDy Manual &raquo;</a></li>
          <li><a href="../index.html" >Reference Manual &raquo;</a></li>
          <li><a href="index.html" accesskey="U">Sequence Analysis &raquo;</a></li> 
        </ul>
      </div>
    </div>
  </div>

  <div class="content">
  <div class="row">
    <div class="span3">
      <div id="sidebar" style="width: 220px">
        <div class="accordion" id="accordion2">
  <div id="searchbox" class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseToolbox">
        <strong>Toolbox</strong>
      </a>
    </div>
    <div id="collapseToolbox" class="accordion-body collapse in">
      <div class="accordion-inner">
        <div id="showcodebuttons" style="margin-bottom: 10px">
          <p><small>Show code snippets for copying:</small></p>
          <div class="btn-group">
            <a href="#codemodal" role="button" data-toggle="modal" id="showcode" class="btn btn-primary btn-small">
              <i class="icon-fullscreen icon-white"></i> Show code</a>
            <a href="#codemodal" role="button" data-toggle="modal" id="withcomments" class="btn btn-primary btn-small">
              and output <i class="icon-comment icon-white"></i> </a>
          </div>
        </div>

        <p><small>Search terms or a module, class or function by name:</small></p>
        <form class="form-search pull-right" action="../../../search.html" method="get">
          <div class="input-append">
            <input type="text" class="input-small search-query" name="q" class="span2">
            <button type="submit" class="btn btn-primary">Go</button>
          </div>
          <input type="hidden" name="check_keywords" value="yes" />
          <input type="hidden" name="area" value="default" />
        </form>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseReleaseNotes">
        <strong>Release Notes</strong>
      </a>
    </div>
    <div id="collapseReleaseNotes" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small><strong>v1</strong> series come with new and improved sequence, structure, and dynamics analysis
features. See <a href="../../release/index.html">release notes</a> for details.
        </small></p>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseHowToCite">
        <strong>How to Cite</strong>
      </a>
    </div>
    <div id="collapseHowToCite" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small>
          Bakan A, Meireles LM, Bahar I
          <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/27/11/1575" target="_blank">
          <i>ProDy</i>: Protein Dynamics Inferred from Theory and Experiments</a> <br/>
          <em>Bioinformatics</em> <strong>2011</strong> 27(11):1575-1577.
        </small></p>
      </div>
    </div>
  </div>
        </div>
      </div>
    </div>
    <div class="span9 body">
    
  <div class="section" id="module-prody.sequence.analysis">
<span id="analysis-functions"></span><h1>Analysis Functions<a class="headerlink" href="#module-prody.sequence.analysis" title="Permalink to this headline">¶</a></h1>
<p>This module defines MSA analysis functions.</p>
<dl class="function">
<dt id="prody.sequence.analysis.calcShannonEntropy">
<code class="descname">calcShannonEntropy</code><span class="sig-paren">(</span><em>msa</em>, <em>ambiguity=True</em>, <em>omitgaps=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#calcShannonEntropy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.calcShannonEntropy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns Shannon entropy array calculated for <em>msa</em>, which may be
an <a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> instance or a 2D Numpy character array.  Implementation
is case insensitive and handles ambiguous amino acids as follows:</p>
<blockquote>
<div><ul class="simple">
<li><strong>B</strong> (Asx) count is allocated to <em>D</em> (Asp) and <em>N</em> (Asn)</li>
<li><strong>Z</strong> (Glx) count is allocated to <em>E</em> (Glu) and <em>Q</em> (Gln)</li>
<li><strong>J</strong> (Xle) count is allocated to <em>I</em> (Ile) and <em>L</em> (Leu)</li>
<li><strong>X</strong> (Xaa) count is allocated to the twenty standard amino acids</li>
</ul>
</div></blockquote>
<p>Selenocysteine (<strong>U</strong>, Sec) and pyrrolysine (<strong>O</strong>, Pyl) are considered
as distinct amino acids.  When <em>ambiguity</em> is set <strong>False</strong>, all alphabet
characters as considered as distinct types.</p>
<p>All non-alphabet characters are considered as gaps, and they are handled
in two ways:</p>
<blockquote>
<div><ul class="simple">
<li>non-existent, the probability of observing amino acids in a given
column is adjusted, by default</li>
<li>as a distinct character with its own probability, when <em>omitgaps</em> is
<strong>False</strong></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.buildMutinfoMatrix">
<code class="descname">buildMutinfoMatrix</code><span class="sig-paren">(</span><em>msa</em>, <em>ambiguity=True</em>, <em>turbo=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#buildMutinfoMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.buildMutinfoMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns mutual information matrix calculated for <em>msa</em>, which may be an
<a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> instance or a 2D Numpy character array.  Implementation
is case insensitive and handles ambiguous amino acids as follows:</p>
<blockquote>
<div><ul class="simple">
<li><strong>B</strong> (Asx) count is allocated to <em>D</em> (Asp) and <em>N</em> (Asn)</li>
<li><strong>Z</strong> (Glx) count is allocated to <em>E</em> (Glu) and <em>Q</em> (Gln)</li>
<li><strong>J</strong> (Xle) count is allocated to <em>I</em> (Ile) and <em>L</em> (Leu)</li>
<li><strong>X</strong> (Xaa) count is allocated to the twenty standard amino acids</li>
<li>Joint probability of observing a pair of ambiguous amino acids is
allocated to all potential combinations, e.g. probability of <strong>XX</strong>
is allocated to 400 combinations of standard amino acids, similarly
probability of <strong>XB</strong> is allocated to 40 combinations of <em>D</em> and <em>N</em>
with the standard amino acids.</li>
</ul>
</div></blockquote>
<p>Selenocysteine (<strong>U</strong>, Sec) and pyrrolysine (<strong>O</strong>, Pyl) are considered
as distinct amino acids.  When <em>ambiguity</em> is set <strong>False</strong>, all alphabet
characters as considered as distinct types.  All non-alphabet characters
are considered as gaps.</p>
<p>Mutual information matrix can be normalized or corrected using
<code class="xref py py-func docutils literal"><span class="pre">applyMINormalization()</span></code> and <code class="xref py py-func docutils literal"><span class="pre">applyMICorrection()</span></code> methods,
respectively.  Normalization by joint entropy can performed using this
function with <em>norm</em> option set <strong>True</strong>.</p>
<p>By default, <em>turbo</em> mode, which uses memory as large as the MSA array
itself but runs four to five times faster, will be used.  If memory
allocation fails, the implementation will fall back to slower and
memory efficient mode.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.calcMSAOccupancy">
<code class="descname">calcMSAOccupancy</code><span class="sig-paren">(</span><em>msa</em>, <em>occ='res'</em>, <em>count=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#calcMSAOccupancy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.calcMSAOccupancy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns occupancy array calculated for residue positions (default,
<code class="docutils literal"><span class="pre">'res'</span></code> or <code class="docutils literal"><span class="pre">'col'</span></code> for <em>occ</em>) or sequences (<code class="docutils literal"><span class="pre">'seq'</span></code> or <code class="docutils literal"><span class="pre">'row'</span></code>
for <em>occ</em>) of <em>msa</em>, which may be an <a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> instance or a 2D
NumPy character array.  By default, occupancy [0-1] will be calculated.
If <em>count</em> is <strong>True</strong>, count of non-gap characters will be returned.
Implementation is case insensitive.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.applyMutinfoCorr">
<code class="descname">applyMutinfoCorr</code><span class="sig-paren">(</span><em>mutinfo</em>, <em>corr='prod'</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#applyMutinfoCorr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.applyMutinfoCorr" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of <em>mutinfo</em> array after average product correction
(default) or average sum correction is applied.  See <a class="reference internal" href="#dsd08" id="id1">[DSD08]</a> for details.</p>
<table class="docutils citation" frame="void" id="dsd08" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[DSD08]</a></td><td>Dunn SD, Wahl LM, Gloor GB. Mutual information without the
influence of phylogeny or entropy dramatically improves residue
contact prediction. <em>Bioinformatics</em> <strong>2008</strong> 24(3):333-340.</td></tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.applyMutinfoNorm">
<code class="descname">applyMutinfoNorm</code><span class="sig-paren">(</span><em>mutinfo</em>, <em>entropy</em>, <em>norm='sument'</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#applyMutinfoNorm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.applyMutinfoNorm" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply one of the normalizations discussed in <a class="reference internal" href="#mlc05" id="id2">[MLC05]</a> to <em>mutinfo</em>
matrix.  <em>norm</em> can be one of the following:</p>
<blockquote>
<div><ul>
<li><p class="first"><code class="docutils literal"><span class="pre">'sument'</span></code>: <img src="../../../_images/mathmpl/math-3044c0900e.png" style="position: relative; bottom: -8px"/>, sum of entropy of columns</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">'minent'</span></code>: <img src="../../../_images/mathmpl/math-0b2e061dbe.png" style="position: relative; bottom: -8px"/>, minimum entropy</p>
</li>
<li><p class="first"><code class="docutils literal"><span class="pre">'maxent'</span></code>: <img src="../../../_images/mathmpl/math-65292337c7.png" style="position: relative; bottom: -8px"/>, maximum entropy</p>
</li>
<li><dl class="first docutils">
<dt><code class="docutils literal"><span class="pre">'mincon'</span></code>: <img src="../../../_images/mathmpl/math-5336060960.png" style="position: relative; bottom: -8px"/>, minimum conditional</dt>
<dd><p class="first last">entropy</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><code class="docutils literal"><span class="pre">'maxcon'</span></code>: <img src="../../../_images/mathmpl/math-8ae26543d3.png" style="position: relative; bottom: -8px"/>, maximum conditional</dt>
<dd><p class="first last">entropy</p>
</dd>
</dl>
</li>
</ul>
</div></blockquote>
<p>where <img src="../../../_images/mathmpl/math-6b8900c6fc.png" style="position: relative; bottom: -8px"/> is the entropy of a column, and
<img src="../../../_images/mathmpl/math-2718240489.png" style="position: relative; bottom: -8px"/>.  Normalization with joint entropy, i.e.
<img src="../../../_images/mathmpl/math-6f75d84fd2.png" style="position: relative; bottom: -8px"/>, can be done using <a class="reference internal" href="#prody.sequence.analysis.buildMutinfoMatrix" title="prody.sequence.analysis.buildMutinfoMatrix"><code class="xref py py-func docutils literal"><span class="pre">buildMutinfoMatrix()</span></code></a> <em>norm</em>
argument.</p>
<table class="docutils citation" frame="void" id="mlc05" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[MLC05]</a></td><td>Martin LC, Gloor GB, Dunn SD, Wahl LM. Using information theory
to search for co-evolving residues in proteins. <em>Bioinformatics</em>
<strong>2005</strong> 21(22):4116-4124.</td></tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.calcRankorder">
<code class="descname">calcRankorder</code><span class="sig-paren">(</span><em>matrix</em>, <em>zscore=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#calcRankorder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.calcRankorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Returnss indices of elements and corresponding values sorted in
descending order, if <em>descend</em> is <strong>True</strong> (default). Can apply a zscore
normalization; by default along <em>axis</em> - 0 such that each column has
mean=0 and std=1.  If <em>zcore</em> analysis is used, return value contains the
zscores. If matrix is smymetric only lower triangle indices will be
returned, with diagonal elements if <em>diag</em> is <strong>True</strong> (default).</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.buildSeqidMatrix">
<code class="descname">buildSeqidMatrix</code><span class="sig-paren">(</span><em>msa</em>, <em>turbo=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#buildSeqidMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.buildSeqidMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns sequence identity matrix for <em>msa</em>.</p>
<p>By default, <em>turbo</em> mode, which uses memory as large as the MSA array
itself but runs four to five times faster, will be used.  If memory
allocation fails, the implementation will fall back to slower and
memory efficient mode.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.uniqueSequences">
<code class="descname">uniqueSequences</code><span class="sig-paren">(</span><em>msa</em>, <em>seqid=0.98</em>, <em>turbo=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#uniqueSequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.uniqueSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a boolean array marking unique sequences in <em>msa</em>.  A sequence
sharing sequence identity of <em>sqid</em> or more with another sequence coming
before itself in <em>msa</em> will have a <strong>False</strong> value in the array.</p>
<p>By default, <em>turbo</em> mode, which uses memory as large as the MSA array
itself but runs four to five times faster, will be used.  If memory
allocation fails, the implementation will fall back to slower and
memory efficient mode.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.buildOMESMatrix">
<code class="descname">buildOMESMatrix</code><span class="sig-paren">(</span><em>msa</em>, <em>ambiguity=True</em>, <em>turbo=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#buildOMESMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.buildOMESMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns OMES (Observed Minus Expected Squared) covariance matrix
calculated for <em>msa</em>, which may be an <a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> instance or a 2D
NumPy character array. OMES is defined as:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>                  (N_OBS - N_EX)^2              (f_i,j - f_i * f_j)^2
OMES_(i,j) = sum(------------------) = N * sum(-----------------------)
                       N_EX                           f_i * f_j
</pre></div>
</div>
<p>Implementation is case insensitive and handles ambiguous amino acids
as follows:</p>
<blockquote>
<div><ul class="simple">
<li><strong>B</strong> (Asx) count is allocated to <em>D</em> (Asp) and <em>N</em> (Asn)</li>
<li><strong>Z</strong> (Glx) count is allocated to <em>E</em> (Glu) and <em>Q</em> (Gln)</li>
<li><strong>J</strong> (Xle) count is allocated to <em>I</em> (Ile) and <em>L</em> (Leu)</li>
<li><strong>X</strong> (Xaa) count is allocated to the twenty standard amino acids</li>
<li>Joint probability of observing a pair of ambiguous amino acids is
allocated to all potential combinations, e.g. probability of <strong>XX</strong>
is allocated to 400 combinations of standard amino acids, similarly
probability of <strong>XB</strong> is allocated to 40 combinations of <em>D</em> and <em>N</em>
with the standard amino acids.</li>
</ul>
</div></blockquote>
<p>Selenocysteine (<strong>U</strong>, Sec) and pyrrolysine (<strong>O</strong>, Pyl) are considered
as distinct amino acids.  When <em>ambiguity</em> is set <strong>False</strong>, all alphabet
characters as considered as distinct types.  All non-alphabet characters
are considered as gaps.</p>
<p>By default, <em>turbo</em> mode, which uses memory as large as the MSA array
itself but runs four to five times faster, will be used.  If memory
allocation fails, the implementation will fall back to slower and
memory efficient mode.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.buildSCAMatrix">
<code class="descname">buildSCAMatrix</code><span class="sig-paren">(</span><em>msa</em>, <em>turbo=True</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#buildSCAMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.buildSCAMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns SCA matrix calculated for <em>msa</em>, which may be an <a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a>
instance or a 2D Numpy character array.</p>
<p>Implementation is case insensitive and handles ambiguous amino acids
as follows:</p>
<blockquote>
<div><ul class="simple">
<li><strong>B</strong> (Asx) count is allocated to <em>D</em> (Asp) and <em>N</em> (Asn)</li>
<li><strong>Z</strong> (Glx) count is allocated to <em>E</em> (Glu) and <em>Q</em> (Gln)</li>
<li><strong>J</strong> (Xle) count is allocated to <em>I</em> (Ile) and <em>L</em> (Leu)</li>
<li><strong>X</strong> (Xaa) count is allocated to the twenty standard amino acids</li>
<li>Joint probability of observing a pair of ambiguous amino acids is
allocated to all potential combinations, e.g. probability of <strong>XX</strong>
is allocated to 400 combinations of standard amino acids, similarly
probability of <strong>XB</strong> is allocated to 40 combinations of <em>D</em> and <em>N</em>
with the standard amino acids.</li>
</ul>
</div></blockquote>
<p>Selenocysteine (<strong>U</strong>, Sec) and pyrrolysine (<strong>O</strong>, Pyl) are considered
as distinct amino acids.  When <em>ambiguity</em> is set <strong>False</strong>, all alphabet
characters as considered as distinct types.  All non-alphabet characters
are considered as gaps.</p>
<p>By default, <em>turbo</em> mode, which uses memory as large as the MSA array
itself but runs four to five times faster, will be used.  If memory
allocation fails, the implementation will fall back to slower and
memory efficient mode.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.buildDirectInfoMatrix">
<code class="descname">buildDirectInfoMatrix</code><span class="sig-paren">(</span><em>msa</em>, <em>seqid=0.8</em>, <em>pseudo_weight=0.5</em>, <em>refine=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#buildDirectInfoMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.buildDirectInfoMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns direct information matrix calculated for <em>msa</em>, which may be an
<a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> instance or a 2D Numpy character array.</p>
<p>Sequences sharing sequence identity of <em>seqid</em> or more with another
sequence are regarded as similar sequences for calculating their weights
using <a class="reference internal" href="#prody.sequence.analysis.calcMeff" title="prody.sequence.analysis.calcMeff"><code class="xref py py-func docutils literal"><span class="pre">calcMeff()</span></code></a>.</p>
<p><em>pseudo_weight</em> are the weight for pseudo count probability.</p>
<p>Sequences are not refined by default. When <em>refine</em> is set <strong>True</strong>,
the MSA will be refined by the first sequence and the shape of direct
information matrix will be smaller.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.calcMeff">
<code class="descname">calcMeff</code><span class="sig-paren">(</span><em>msa</em>, <em>seqid=0.8</em>, <em>refine=False</em>, <em>weight=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#calcMeff"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.calcMeff" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the Meff for <em>msa</em>, which may be an <a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a>
instance or a 2D Numpy character array.</p>
<p>Since similar sequences in an <em>msa</em> decreases the diversity of <em>msa</em>,
<em>Meff</em> gives a weight for sequences in the <em>msa</em>.</p>
<p>For example: One sequence in MSA has 5 other similar sequences in this
MSA(itself included). The weight of this sequence is defined as 1/5=0.2.
Meff is the sum of all sequence weights. In another word, Meff can be
understood as the effective number of independent sequences.</p>
<p>Sequences sharing sequence identity of <em>seqid</em> or more with another
sequence are regarded as similar sequences to calculate Meff.</p>
<p>Sequences are not refined by default. When <em>refine</em> is set <strong>True</strong>, the
MSA will be refined by the first sequence.</p>
<p>The weight for each sequence are returned when <em>weight</em> is <strong>True</strong>.</p>
</dd></dl>

<dl class="function">
<dt id="prody.sequence.analysis.buildPCMatrix">
<code class="descname">buildPCMatrix</code><span class="sig-paren">(</span><em>msa</em>, <em>turbo=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/prody/sequence/analysis.html#buildPCMatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#prody.sequence.analysis.buildPCMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns PC matrix calculated for <em>msa</em>, which may be an <a class="reference internal" href="msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a>
instance or a 2D Numpy character array.</p>
<p>Implementation is case insensitive and handles ambiguous amino acids
as follows:</p>
<blockquote>
<div><ul class="simple">
<li><strong>B</strong> (Asx) count is allocated to <em>D</em> (Asp) and <em>N</em> (Asn)</li>
<li><strong>Z</strong> (Glx) count is allocated to <em>E</em> (Glu) and <em>Q</em> (Gln)</li>
<li><strong>J</strong> (Xle) count is allocated to <em>I</em> (Ile) and <em>L</em> (Leu)</li>
<li><strong>X</strong> (Xaa) count is allocated to the twenty standard amino acids</li>
<li>Joint probability of observing a pair of ambiguous amino acids is
allocated to all potential combinations, e.g. probability of <strong>XX</strong>
is allocated to 400 combinations of standard amino acids, similarly
probability of <strong>XB</strong> is allocated to 40 combinations of <em>D</em> and <em>N</em>
with the standard amino acids.</li>
</ul>
</div></blockquote>
<p>Selenocysteine (<strong>U</strong>, Sec) and pyrrolysine (<strong>O</strong>, Pyl) are considered
as distinct amino acids.  When <em>ambiguity</em> is set <strong>False</strong>, all alphabet
characters as considered as distinct types.  All non-alphabet characters
are considered as gaps.</p>
</dd></dl>

</div>


    </div>
  </div>
</div>
<footer class="footer">
  <div class="row">
    <div class="span12">
      <p><small>
      &copy; Copyright 2010-2017, University of Pittsburgh.
      Last updated on Dec 06, 2017.
      Created using <a href="http://sphinx-doc.org">Sphinx</a> 1.3.5.
      </small></p>
    </div>
  </div>
  </div>
</footer>
<div id="codemodal" class="modal hide fade" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
  <div class="modal-header">
    <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
    <h3 id="myModalLabel">Code Snippets</h3>
  </div>
  <div class="modal-body">
    <textarea id="codesnippets"></textarea>
  </div>
  <div class="modal-footer">
    <button class="btn btn-info" id="selectcode">Select Code</button>
    <!--
    <button class="btn btn-success" id="copycode" data-clipboard-target="codesnippets">Copy Code</button>
    -->
    <button class="btn" data-dismiss="modal" aria-hidden="true">Close</button>
  </div>
  <input type="hidden" id="zeroclipboardpath" value="../../../_static/js/zeroclipboard.swf">
</div>
  </body>
</html>