
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">



<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-19801227-1']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
    <title>ProDy 1.4 Series &mdash; RuleBender</title>
    <link rel="stylesheet" href="../../../_static/css/basic.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/css/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/css/bootswatch-united.min.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/css/prodydocs.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  false
      };
    </script>
    <script type="text/javascript" src="../../../_static/js/jquery-2.0.0.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/zeroclipboard.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/copybutton.js"></script>
    <script type="text/javascript" src="../../../_static/js/prodydocs.js"></script>
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>

  </head>
  <body data-spy="scroll" data-target="#localtoc" data-offset="190">

<div class="container">
  <div class="row">
    <div class="span12">
      <a href="../../../index.html">
        <img src="../../../_static/logo.png" alt="ProDy logo" id="logo" />
      </a>
    </div>
    <div class="span12">
      <div class="subnav">
        <ul class="nav nav-pills">
        <li class="pull-right">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a>
          </li>
        <li class="pull-right">
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a>
          </li>
        <li><a href="../../../index.html">Home  &raquo;</a></li> 
        </ul>
      </div>
    </div>
  </div>

  <div class="content">
  <div class="row">
    <div class="span3">
      <div id="sidebar" style="width: 220px">
  <div id="localtoc" style="margin-bottom: 20px">
    <ul>
<li><a class="reference internal" href="#">ProDy 1.4 Series</a><ul>
<li><a class="reference internal" href="#nov-14-2013">1.4.9 (Nov 14, 2013)</a></li>
<li><a class="reference internal" href="#nov-4-2013">1.4.8 (Nov 4, 2013)</a></li>
<li><a class="reference internal" href="#oct-29-2013">1.4.7 (Oct 29, 2013)</a></li>
<li><a class="reference internal" href="#oct-16-2013">1.4.6 (Oct 16, 2013)</a></li>
<li><a class="reference internal" href="#sep-6-2013">1.4.5 (Sep 6, 2013)</a></li>
<li><a class="reference internal" href="#july-22-2013">1.4.4 (July 22, 2013)</a></li>
<li><a class="reference internal" href="#june-14-2013">1.4.3 (June 14, 2013)</a></li>
<li><a class="reference internal" href="#april-19-2013">1.4.2 (April 19, 2013)</a></li>
<li><a class="reference internal" href="#dec-16-2012">1.4.1 (Dec 16, 2012)</a><ul>
<li><a class="reference internal" href="#normal-mode-wizard">Normal Mode Wizard</a></li>
</ul>
</li>
<li><a class="reference internal" href="#dec-2-2012">1.4 (Dec 2, 2012)</a></li>
</ul>
</li>
</ul>

  </div>
        <div class="accordion" id="accordion2">
  <div id="searchbox" class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseToolbox">
        <strong>Toolbox</strong>
      </a>
    </div>
    <div id="collapseToolbox" class="accordion-body collapse in">
      <div class="accordion-inner">
        <div id="showcodebuttons" style="margin-bottom: 10px">
          <p><small>Show code snippets for copying:</small></p>
          <div class="btn-group">
            <a href="#codemodal" role="button" data-toggle="modal" id="showcode" class="btn btn-primary btn-small">
              <i class="icon-fullscreen icon-white"></i> Show code</a>
            <a href="#codemodal" role="button" data-toggle="modal" id="withcomments" class="btn btn-primary btn-small">
              and output <i class="icon-comment icon-white"></i> </a>
          </div>
        </div>

        <p><small>Search terms or a module, class or function by name:</small></p>
        <form class="form-search pull-right" action="../../../search.html" method="get">
          <div class="input-append">
            <input type="text" class="input-small search-query" name="q" class="span2">
            <button type="submit" class="btn btn-primary">Go</button>
          </div>
          <input type="hidden" name="check_keywords" value="yes" />
          <input type="hidden" name="area" value="default" />
        </form>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseReleaseNotes">
        <strong>Release Notes</strong>
      </a>
    </div>
    <div id="collapseReleaseNotes" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small><strong>v1</strong> series come with new and improved sequence, structure, and dynamics analysis
features. See <a href="../../../manual/release/index.html">release notes</a> for details.
        </small></p>
      </div>
    </div>
  </div>  <div class="accordion-group">
    <div class="accordion-heading">
      <a class="accordion-toggle" data-toggle="collapse" data-parent="#accordion2" href="#collapseHowToCite">
        <strong>How to Cite</strong>
      </a>
    </div>
    <div id="collapseHowToCite" class="accordion-body collapse in">
      <div class="accordion-inner">
        <p><small>
          Bakan A, Meireles LM, Bahar I
          <a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/27/11/1575" target="_blank">
          <i>ProDy</i>: Protein Dynamics Inferred from Theory and Experiments</a> <br/>
          <em>Bioinformatics</em> <strong>2011</strong> 27(11):1575-1577.
        </small></p>
      </div>
    </div>
  </div>
        </div>
      </div>
    </div>
    <div class="span9 body">
    
  <div class="section" id="prody-1-4-series">
<h1>ProDy 1.4 Series<a class="headerlink" href="#prody-1-4-series" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><a class="reference internal" href="#nov-14-2013" id="id1">1.4.9 (Nov 14, 2013)</a></li>
<li><a class="reference internal" href="#nov-4-2013" id="id2">1.4.8 (Nov 4, 2013)</a></li>
<li><a class="reference internal" href="#oct-29-2013" id="id3">1.4.7 (Oct 29, 2013)</a></li>
<li><a class="reference internal" href="#oct-16-2013" id="id4">1.4.6 (Oct 16, 2013)</a></li>
<li><a class="reference internal" href="#sep-6-2013" id="id5">1.4.5 (Sep 6, 2013)</a></li>
<li><a class="reference internal" href="#july-22-2013" id="id6">1.4.4 (July 22, 2013)</a></li>
<li><a class="reference internal" href="#june-14-2013" id="id7">1.4.3 (June 14, 2013)</a></li>
<li><a class="reference internal" href="#april-19-2013" id="id8">1.4.2 (April 19, 2013)</a></li>
<li><a class="reference internal" href="#dec-16-2012" id="id9">1.4.1 (Dec 16, 2012)</a><ul>
<li><a class="reference internal" href="#normal-mode-wizard" id="id10">Normal Mode Wizard</a></li>
</ul>
</li>
<li><a class="reference internal" href="#dec-2-2012" id="id11">1.4 (Dec 2, 2012)</a></li>
</ul>
</div>
<div class="section" id="nov-14-2013">
<h2><a class="toc-backref" href="#id1">1.4.9 (Nov 14, 2013)</a><a class="headerlink" href="#nov-14-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>Upcoming changes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Support for Python 3.1 and NumPy 1.5 will be dropped, meaning no Windows
installers will be built for these versions of them.</li>
</ul>
</div></blockquote>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/atomic/hierview.html#prody.atomic.hierview.HierView" title="prody.atomic.hierview.HierView"><code class="xref py py-class docutils literal"><span class="pre">HierView</span></code></a> can handle <a class="reference internal" href="../../../manual/reference/atomic/residue.html#prody.atomic.residue.Residue" title="prody.atomic.residue.Residue"><code class="xref py py-class docutils literal"><span class="pre">Residue</span></code></a> instances that have
same <a class="reference internal" href="../../../manual/reference/atomic/fields.html#term-segment"><span class="xref std std-term">segment</span></a> name, <a class="reference internal" href="../../../manual/reference/atomic/fields.html#term-chain"><span class="xref std std-term">chain</span></a> identifier, and <a class="reference internal" href="../../../manual/reference/atomic/fields.html#term-resnum"><span class="xref std std-term">resnum</span></a>,
if PDB file contains <code class="docutils literal"><span class="pre">TER</span></code> lines to terminate these residues. If
these three identifiers are shared by multiple residues, indexing
<a class="reference internal" href="../../../manual/reference/atomic/atomgroup.html#prody.atomic.atomgroup.AtomGroup" title="prody.atomic.atomgroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a> instances will return a list of residues. This
behavior can be used as follows. Note that in v1.5, this will be the
default behavior.</li>
</ul>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pdb_lines</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="gp">... </span><span class="s2">ATOM      1  O   WAT A   1       4.694  -3.891  -0.592  1.00  1.00</span>
<span class="gp">... </span><span class="s2">ATOM      2  H1  WAT A   1       5.096  -3.068  -0.190  1.00  1.00</span>
<span class="gp">... </span><span class="s2">ATOM      3  H2  WAT A   1       5.420  -4.544  -0.808  1.00  1.00</span>
<span class="gp">... </span><span class="s2">TER</span>
<span class="gp">... </span><span class="s2">ATOM      4  O   WAT A   1     -30.035  19.116  -2.193  1.00  1.00</span>
<span class="gp">... </span><span class="s2">ATOM      5  H1  WAT A   1     -30.959  18.736  -2.244  1.00  1.00</span>
<span class="gp">... </span><span class="s2">ATOM      6  H2  WAT A   1     -29.993  19.960  -2.728  1.00  1.00</span>
<span class="gp">... </span><span class="s2">TER</span>
<span class="gp">... </span><span class="s2">ATOM      7  O   WAT A   1     -77.584 -21.524 -37.894  1.00  1.00</span>
<span class="gp">... </span><span class="s2">ATOM      8  H1  WAT A   1     -77.226 -21.966 -38.717  1.00  1.00</span>
<span class="gp">... </span><span class="s2">ATOM      9  H2  WAT A   1     -77.023 -20.726 -37.674  1.00  1.00</span>
<span class="gp">... </span><span class="s2">TER&quot;&quot;&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">StringIO</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">atoms</span> <span class="o">=</span> <span class="n">parsePDBStream</span><span class="p">(</span><span class="n">StringIO</span><span class="p">(</span><span class="n">pdb_lines</span><span class="p">))</span>
</pre></div>
</div>
<blockquote>
<div>Current behavior:</div></blockquote>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">atoms</span><span class="o">.</span><span class="n">numResidues</span><span class="p">())</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">atoms</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="go">&lt;Residue: WAT 1 from Chain A from Unknown (9 atoms)&gt;</span>
</pre></div>
</div>
<blockquote>
<div>To activate the new behavior (which will be the default behavior in v1.5):</div></blockquote>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hv</span> <span class="o">=</span> <span class="n">atoms</span><span class="o">.</span><span class="n">getHierView</span><span class="p">(</span><span class="n">ter</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">hv</span><span class="o">.</span><span class="n">numResidues</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hv</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/proteins/pdbfile.html#prody.proteins.pdbfile.parsePDB" title="prody.proteins.pdbfile.parsePDB"><code class="xref py py-func docutils literal"><span class="pre">parsePDB()</span></code></a> reads <code class="docutils literal"><span class="pre">TER</span></code> records in PDB files. Atoms and hetero
atoms (<a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-hetatm"><span class="xref std std-term">hetatm</span></a>) that are followed by a <code class="docutils literal"><span class="pre">TER</span></code> record are now
flagged as <a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-pdbter"><span class="xref std std-term">pdbter</span></a>.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Fixed memory leaks in <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.uniqueSequences" title="prody.sequence.analysis.uniqueSequences"><code class="xref py py-func docutils literal"><span class="pre">uniqueSequences()</span></code></a> and
<a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.buildSeqidMatrix" title="prody.sequence.analysis.buildSeqidMatrix"><code class="xref py py-func docutils literal"><span class="pre">buildSeqidMatrix()</span></code></a>.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="nov-4-2013">
<h2><a class="toc-backref" href="#id2">1.4.8 (Nov 4, 2013)</a><a class="headerlink" href="#nov-4-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>New Features</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>New analysis functions <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.buildOMESMatrix" title="prody.sequence.analysis.buildOMESMatrix"><code class="xref py py-func docutils literal"><span class="pre">buildOMESMatrix()</span></code></a> and <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.buildSCAMatrix" title="prody.sequence.analysis.buildSCAMatrix"><code class="xref py py-func docutils literal"><span class="pre">buildSCAMatrix()</span></code></a>
are implemented.</li>
<li>New <a class="reference internal" href="../../../manual/reference/atomic/atomgroup.html#prody.atomic.atomgroup.AtomGroup.numBytes" title="prody.atomic.atomgroup.AtomGroup.numBytes"><code class="xref py py-meth docutils literal"><span class="pre">AtomGroup.numBytes()</span></code></a> method returns an estimate of memory usage.</li>
<li>New <a class="reference internal" href="../../../manual/reference/utilities/misctools.html#prody.utilities.misctools.countBytes" title="prody.utilities.misctools.countBytes"><code class="xref py py-func docutils literal"><span class="pre">countBytes()</span></code></a> utility function is added for counting bytes used
by NumPy arrays.</li>
</ul>
</div></blockquote>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/proteins/pdbfile.html#prody.proteins.pdbfile.parsePDB" title="prody.proteins.pdbfile.parsePDB"><code class="xref py py-func docutils literal"><span class="pre">parsePDB()</span></code></a> resizes data arrays to decrease memory usage.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Fixed memory leaks in MSA <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#module-prody.sequence.analysis" title="prody.sequence.analysis"><code class="xref py py-mod docutils literal"><span class="pre">analysis</span></code></a> functions.</li>
<li>Fixed potential problems with importing contributed libraries.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="oct-29-2013">
<h2><a class="toc-backref" href="#id3">1.4.7 (Oct 29, 2013)</a><a class="headerlink" href="#oct-29-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/atomic/atomgroup.html#prody.atomic.atomgroup.AtomGroup" title="prody.atomic.atomgroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>, <a class="reference internal" href="../../../manual/reference/atomic/selection.html#prody.atomic.selection.Selection" title="prody.atomic.selection.Selection"><code class="xref py py-class docutils literal"><span class="pre">Selection</span></code></a>, and other <a class="reference internal" href="../../../manual/reference/atomic/atomic.html#prody.atomic.atomic.Atomic" title="prody.atomic.atomic.Atomic"><code class="xref py py-class docutils literal"><span class="pre">Atomic</span></code></a>
classes are picklable.</li>
<li>Improved equality tests for <a class="reference internal" href="../../../manual/reference/atomic/atomgroup.html#prody.atomic.atomgroup.AtomGroup" title="prody.atomic.atomgroup.AtomGroup"><code class="xref py py-class docutils literal"><span class="pre">AtomGroup</span></code></a>. Two different instances
are considered equal if they contain identical data and coordinate sets.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="oct-16-2013">
<h2><a class="toc-backref" href="#id4">1.4.6 (Oct 16, 2013)</a><a class="headerlink" href="#oct-16-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Selection problem with using <a class="reference internal" href="../../../manual/reference/atomic/fields.html#term-resid"><span class="xref std std-term">resid</span></a> is fixed (<a class="reference external" href="https://github.com/prody/ProDy/issues/160">issue 160</a>)</li>
<li>Fixed a memory leak in MSA parsers written in C.  When dealing with
large files, leak would cause a segmentation fault.</li>
<li>Fixed a memory leak in MSA parsers written in C.  When dealing with
large files, leak would cause a segmentation fault.</li>
<li>Fixed a reference counting problem in MSA parsers in C that would
cause segmentation fault when reading files that uses the same
label for multiple sequences.</li>
<li>Updated <a class="reference internal" href="../../../manual/reference/proteins/pdbligands.html#prody.proteins.pdbligands.fetchPDBLigand" title="prody.proteins.pdbligands.fetchPDBLigand"><code class="xref py py-func docutils literal"><span class="pre">fetchPDBLigand()</span></code></a> to use PDB for fetching XML files.</li>
<li>Revised handling of MSA file formats to avoid exceptions for unknown
extensions.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="sep-6-2013">
<h2><a class="toc-backref" href="#id5">1.4.5 (Sep 6, 2013)</a><a class="headerlink" href="#sep-6-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>New Features</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/proteins/header.html#prody.proteins.header.parsePDBHeader" title="prody.proteins.header.parsePDBHeader"><code class="xref py py-func docutils literal"><span class="pre">parsePDBHeader()</span></code></a> function can parse space group information
from header section specified as <code class="docutils literal"><span class="pre">REMARK</span> <span class="pre">290</span></code>, e.g.
<code class="docutils literal"><span class="pre">parsePDBHeader('1mkp',</span> <span class="pre">'space_group')</span></code> or
<code class="docutils literal"><span class="pre">parsePDBHeader('1mkp')['space_group']</span></code></li>
<li><a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-heavy"><span class="xref std std-term">heavy</span></a> selection flag is defined as an alias for <a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-noh"><span class="xref std std-term">noh</span></a>.</li>
<li><a class="reference internal" href="../../../manual/reference/proteins/compare.html#prody.proteins.compare.matchChains" title="prody.proteins.compare.matchChains"><code class="xref py py-func docutils literal"><span class="pre">matchChains()</span></code></a> function can match non-hydrogen atoms using
<code class="docutils literal"><span class="pre">subset='heavy'</span></code> keyword argument.</li>
<li>Added <code class="docutils literal"><span class="pre">update_coords</span></code> keyword argument to <code class="xref py py-meth docutils literal"><span class="pre">PCA.builCovariance()</span></code>,
so that average coordinates calculated internally can be stored in
ensemble or trajectory objects used as input.</li>
</ul>
</div></blockquote>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Unit tests can be run with Python 2.6 when <cite>unittest2</cite> module is installed.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Fixed problems with reading compressed PDB files using Python 3.3.</li>
<li>Fixed a bug in <a class="reference internal" href="../../../manual/reference/proteins/stride.html#prody.proteins.stride.parseSTRIDE" title="prody.proteins.stride.parseSTRIDE"><code class="xref py py-func docutils literal"><span class="pre">parseSTRIDE()</span></code></a> function that prevented reading files.</li>
<li>Improved parsing of biomolecular transformations.</li>
<li>Fixed memory allocation in C code used by <a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.parseMSA" title="prody.sequence.msafile.parseMSA"><code class="xref py py-func docutils literal"><span class="pre">parseMSA()</span></code></a> (Python 2.6).</li>
<li>Fixed a potential name error in trajectory classes.</li>
<li>Fixed problems in handling compressed files when using Python 2.6 and 3.3.</li>
<li>Fixed a problem with indexing <a class="reference internal" href="../../../manual/reference/dynamics/nma.html#prody.dynamics.nma.NMA" title="prody.dynamics.nma.NMA"><code class="xref py py-class docutils literal"><span class="pre">NMA</span></code></a> instances in Python 3 series.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="july-22-2013">
<h2><a class="toc-backref" href="#id6">1.4.4 (July 22, 2013)</a><a class="headerlink" href="#july-22-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/dynamics/nmdfile.html#prody.dynamics.nmdfile.writeNMD" title="prody.dynamics.nmdfile.writeNMD"><code class="xref py py-func docutils literal"><span class="pre">writeNMD()</span></code></a> and <a class="reference internal" href="../../../manual/reference/dynamics/nmdfile.html#prody.dynamics.nmdfile.parseNMD" title="prody.dynamics.nmdfile.parseNMD"><code class="xref py py-func docutils literal"><span class="pre">parseNMD()</span></code></a> write and read segment names.
NMWiz is also improved to handle segment names. Improvements
will be available in VMD v1.9.2.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>A bug in <a class="reference internal" href="../../../manual/reference/atomic/functions.html#prody.atomic.functions.saveAtoms" title="prody.atomic.functions.saveAtoms"><code class="xref py py-func docutils literal"><span class="pre">saveAtoms()</span></code></a> that would cause <code class="xref py py-exc docutils literal"><span class="pre">KeyError</span></code> when
bonds are set but fragments are not determined is fixed.</li>
<li>Import ProDy would fail when <span class="target" id="index-0"></span><a class="reference external" href="http://matplotlib.sourceforge.net/faq/environment_variables_faq.html#envvar-HOME" title="(in Matplotlib v2.1.0)"><code class="xref std std-envvar docutils literal"><span class="pre">HOME</span></code></a> is not set. Changed
<a class="reference internal" href="../../../manual/reference/utilities/settings.html#prody.utilities.settings.PackageSettings" title="prody.utilities.settings.PackageSettings"><code class="xref py py-class docutils literal"><span class="pre">PackageSettings</span></code></a> to handle this case graciously.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="june-14-2013">
<h2><a class="toc-backref" href="#id7">1.4.3 (June 14, 2013)</a><a class="headerlink" href="#june-14-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>Changes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/dynamics/nmdfile.html#prody.dynamics.nmdfile.getVMDpath" title="prody.dynamics.nmdfile.getVMDpath"><code class="xref py py-func docutils literal"><span class="pre">getVMDpath()</span></code></a> and <a class="reference internal" href="../../../manual/reference/dynamics/nmdfile.html#prody.dynamics.nmdfile.setVMDpath" title="prody.dynamics.nmdfile.setVMDpath"><code class="xref py py-func docutils literal"><span class="pre">setVMDpath()</span></code></a> functions are deprecated for
removal, use <a class="reference internal" href="../../../manual/reference/dynamics/nmdfile.html#prody.dynamics.nmdfile.pathVMD" title="prody.dynamics.nmdfile.pathVMD"><code class="xref py py-func docutils literal"><span class="pre">pathVMD()</span></code></a> instead.</li>
<li>Increased <a class="reference internal" href="../../../manual/reference/proteins/blastpdb.html#prody.proteins.blastpdb.blastPDB" title="prody.proteins.blastpdb.blastPDB"><code class="xref py py-func docutils literal"><span class="pre">blastPDB()</span></code></a> <em>timeout</em> to 60 seconds.</li>
<li><a class="reference internal" href="../../../manual/reference/dynamics/editing.html#prody.dynamics.editing.extendModel" title="prody.dynamics.editing.extendModel"><code class="xref py py-func docutils literal"><span class="pre">extendModel()</span></code></a> and <a class="reference internal" href="../../../manual/reference/dynamics/editing.html#prody.dynamics.editing.extendMode" title="prody.dynamics.editing.extendMode"><code class="xref py py-func docutils literal"><span class="pre">extendMode()</span></code></a> functions have a new option
for normalizing extended mode(s).</li>
<li><a class="reference internal" href="../../../manual/reference/dynamics/sampling.html#prody.dynamics.sampling.sampleModes" title="prody.dynamics.sampling.sampleModes"><code class="xref py py-func docutils literal"><span class="pre">sampleModes()</span></code></a> and <a class="reference internal" href="../../../manual/reference/dynamics/sampling.html#prody.dynamics.sampling.traverseMode" title="prody.dynamics.sampling.traverseMode"><code class="xref py py-func docutils literal"><span class="pre">traverseMode()</span></code></a> automatically normalizes
input modes.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>A bug in <a class="reference internal" href="../../../manual/reference/measure/transform.html#prody.measure.transform.applyTransformation" title="prody.measure.transform.applyTransformation"><code class="xref py py-func docutils literal"><span class="pre">applyTransformation()</span></code></a> is fixed. The function would
interpret some external transformation matrices incorrectly.</li>
<li>A bug in <a class="reference internal" href="../../../manual/reference/proteins/pdbligands.html#prody.proteins.pdbligands.fetchPDBLigand" title="prody.proteins.pdbligands.fetchPDBLigand"><code class="xref py py-func docutils literal"><span class="pre">fetchPDBLigand()</span></code></a> function is fixed.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="april-19-2013">
<h2><a class="toc-backref" href="#id8">1.4.2 (April 19, 2013)</a><a class="headerlink" href="#april-19-2013" title="Permalink to this headline">¶</a></h2>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/proteins/localpdb.html#prody.proteins.localpdb.fetchPDB" title="prody.proteins.localpdb.fetchPDB"><code class="xref py py-func docutils literal"><span class="pre">fetchPDB()</span></code></a> and <a class="reference internal" href="../../../manual/reference/proteins/localpdb.html#prody.proteins.localpdb.fetchPDBfromMirror" title="prody.proteins.localpdb.fetchPDBfromMirror"><code class="xref py py-func docutils literal"><span class="pre">fetchPDBfromMirror()</span></code></a> functions can handle
partial PDB mirrors.  See <a class="reference internal" href="../../../manual/reference/proteins/localpdb.html#prody.proteins.localpdb.pathPDBMirror" title="prody.proteins.localpdb.pathPDBMirror"><code class="xref py py-func docutils literal"><span class="pre">pathPDBMirror()</span></code></a> for setting a mirror path.</li>
</ul>
</div></blockquote>
<p><strong>Changes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference external" href="http://www.pdb.org/pdb/ligand/ligandsummary.do?hetId=MSE">MSE</a> is
included in the definition of non-standard amino acids, i.e.
<a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-nonstdaa"><span class="xref std std-term">nonstdaa</span></a>.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Atom selection problems related to using <a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-all"><span class="xref std std-term">all</span></a> and <a class="reference internal" href="../../../manual/reference/atomic/flags.html#term-none"><span class="xref std std-term">none</span></a> in
composite selections, e.g. <code class="docutils literal"><span class="pre">'calpha</span> <span class="pre">and</span> <span class="pre">all'</span></code>, is fixed by defining these
keywords as <a class="reference internal" href="../../../manual/reference/atomic/flags.html#flags"><span>Atom Flags</span></a>.</li>
<li>Fasta files with sequence labels using multiple pipe characters would
cause C parser (and so <a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.parseMSA" title="prody.sequence.msafile.parseMSA"><code class="xref py py-func docutils literal"><span class="pre">parseMSA()</span></code></a>) to fail.  This issue is fixed
by completely disregarding pipe characters.</li>
<li>Empty chain identifiers for PDB hits would cause a problem in parsing
XML results file and <a class="reference internal" href="../../../manual/reference/proteins/blastpdb.html#prody.proteins.blastpdb.blastPDB" title="prody.proteins.blastpdb.blastPDB"><code class="xref py py-func docutils literal"><span class="pre">blastPDB()</span></code></a> would throw an exception.  This
case is handled by slicing the chain identifier string.</li>
<li>A problem in <a class="reference internal" href="../../../manual/reference/dynamics/nmdfile.html#prody.dynamics.nmdfile.viewNMDinVMD" title="prody.dynamics.nmdfile.viewNMDinVMD"><code class="xref py py-func docutils literal"><span class="pre">viewNMDinVMD()</span></code></a> related to module imports is fixed.</li>
<li>A problem with handling weights in <a class="reference internal" href="../../../manual/reference/ensemble/functions.html#prody.ensemble.functions.loadEnsemble" title="prody.ensemble.functions.loadEnsemble"><code class="xref py py-func docutils literal"><span class="pre">loadEnsemble()</span></code></a> is fixed.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="dec-16-2012">
<h2><a class="toc-backref" href="#id9">1.4.1 (Dec 16, 2012)</a><a class="headerlink" href="#dec-16-2012" title="Permalink to this headline">¶</a></h2>
<p><strong>New Features</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.buildSeqidMatrix" title="prody.sequence.analysis.buildSeqidMatrix"><code class="xref py py-func docutils literal"><span class="pre">buildSeqidMatrix()</span></code></a> and <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.uniqueSequences" title="prody.sequence.analysis.uniqueSequences"><code class="xref py py-func docutils literal"><span class="pre">uniqueSequences()</span></code></a> functions
are implemented for comparing sequences in an <a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> object.</li>
<li><a class="reference internal" href="../../../manual/reference/dynamics/heatmapper.html#prody.dynamics.heatmapper.showHeatmap" title="prody.dynamics.heatmapper.showHeatmap"><code class="xref py py-func docutils literal"><span class="pre">showHeatmap()</span></code></a>, <a class="reference internal" href="../../../manual/reference/dynamics/heatmapper.html#prody.dynamics.heatmapper.parseHeatmap" title="prody.dynamics.heatmapper.parseHeatmap"><code class="xref py py-func docutils literal"><span class="pre">parseHeatmap()</span></code></a>, and <a class="reference internal" href="../../../manual/reference/dynamics/heatmapper.html#prody.dynamics.heatmapper.writeHeatmap" title="prody.dynamics.heatmapper.writeHeatmap"><code class="xref py py-func docutils literal"><span class="pre">writeHeatmap()</span></code></a>
functions are implemented to support VMD plugin <a class="reference external" href="http://www.ks.uiuc.edu/Research/vmd/plugins/heatmapper/">Heat Mapper</a> file format.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/sequence.html#prody.sequence.sequence.Sequence" title="prody.sequence.sequence.Sequence"><code class="xref py py-class docutils literal"><span class="pre">Sequence</span></code></a> is implemented to handle individual sequence records
and point to sequences in <a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> instances.</li>
<li><a class="reference internal" href="../../../manual/apps/evol/occupancy.html#evol-occupancy"><span>evol occupancy</span></a> application is implemented for refined MSA
quality checking purposes.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.mergeMSA" title="prody.sequence.msa.mergeMSA"><code class="xref py py-func docutils literal"><span class="pre">mergeMSA()</span></code></a> function and <a class="reference internal" href="../../../manual/apps/evol/merge.html#evol-merge"><span>evol merge</span></a> application are
implemented for merging Pfam MSA to study multi-domain proteins.</li>
</ul>
</div></blockquote>
<p><strong>Improvements</strong>:</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.refineMSA" title="prody.sequence.msa.refineMSA"><code class="xref py py-func docutils literal"><span class="pre">refineMSA()</span></code></a> function and <a class="reference internal" href="../../../manual/apps/evol/refine.html#evol-refine"><span>evol refine</span></a> application
can perform MSA refinements by removing similar sequences.</li>
<li><a class="reference internal" href="../../../manual/reference/proteins/pdbfile.html#prody.proteins.pdbfile.writePDB" title="prody.proteins.pdbfile.writePDB"><code class="xref py py-func docutils literal"><span class="pre">writePDB()</span></code></a> function takes <em>beta</em> and <em>occupancy</em> arguments
to be outputted in corresponding columns.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> indexing and slicing are revised and improved.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.parseMSA" title="prody.sequence.msafile.parseMSA"><code class="xref py py-func docutils literal"><span class="pre">parseMSA()</span></code></a> is improved to handle indexing of sequences that
have the same label in an MSA file, e.g. domains repeated in a protein.</li>
<li><a class="reference internal" href="../../../manual/apps/prody/anm.html#prody-anm"><span>prody anm</span></a>, <a class="reference internal" href="../../../manual/apps/prody/gnm.html#prody-gnm"><span>prody gnm</span></a>, and <a class="reference internal" href="../../../manual/apps/prody/pca.html#prody-pca"><span>prody pca</span></a> applications
can write heatmap files for visualization using NMWiz and Heatmapper
plugins.</li>
<li>Several improvements made to handling sequence labels in Pfam MSA files.
Files that contain sequence parts with same protein UniProt ID are
handled delicately.</li>
</ul>
</div></blockquote>
<p><strong>Changes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>ProDy will not emit a warning message when a wwPDB server is not set
using <a class="reference internal" href="../../../manual/reference/proteins/wwpdb.html#prody.proteins.wwpdb.wwPDBServer" title="prody.proteins.wwpdb.wwPDBServer"><code class="xref py py-func docutils literal"><span class="pre">wwPDBServer()</span></code></a>, and use the default US server.</li>
<li>Indexing <a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> returns <a class="reference internal" href="../../../manual/reference/sequence/sequence.html#prody.sequence.sequence.Sequence" title="prody.sequence.sequence.Sequence"><code class="xref py py-class docutils literal"><span class="pre">Sequence</span></code></a> instances.</li>
<li>Iterating over <a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> and <a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.MSAFile" title="prody.sequence.msafile.MSAFile"><code class="xref py py-class docutils literal"><span class="pre">MSAFile</span></code></a> yields
<a class="reference internal" href="../../../manual/reference/sequence/sequence.html#prody.sequence.sequence.Sequence" title="prody.sequence.sequence.Sequence"><code class="xref py py-class docutils literal"><span class="pre">Sequence</span></code></a> instances.</li>
</ul>
</div></blockquote>
<p><strong>Bugfixes</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Fixed a syntax problem that prevented running ProDy using Python 2.6.</li>
<li>Fixed <a class="reference internal" href="../../../manual/reference/dynamics/nma.html#prody.dynamics.nma.NMA" title="prody.dynamics.nma.NMA"><code class="xref py py-class docutils literal"><span class="pre">NMA</span></code></a> indexing problem that was introduced in v1.4.</li>
</ul>
</div></blockquote>
<div class="section" id="normal-mode-wizard">
<h3><a class="toc-backref" href="#id10">Normal Mode Wizard</a><a class="headerlink" href="#normal-mode-wizard" title="Permalink to this headline">¶</a></h3>
<blockquote>
<div><ul class="simple">
<li>NMWiz can visualize heatmaps linked to structural view via Heatmapper.
Clicking on the heatmap will highlight atom or residue pairs.</li>
<li>ProDy interface has the option to write and load cross-correlations.</li>
<li>NMWiz can determined whether a model is an extended model. For extended
models plotting mobility has been improved. Only a single value per residue
will be plotted, and clicking on the plot will highlight all of the
residue atoms.</li>
</ul>
</div></blockquote>
</div>
</div>
<div class="section" id="dec-2-2012">
<h2><a class="toc-backref" href="#id11">1.4 (Dec 2, 2012)</a><a class="headerlink" href="#dec-2-2012" title="Permalink to this headline">¶</a></h2>
<p><strong>New Features</strong>:</p>
<p><em>Python 3 Support</em></p>
<blockquote>
<div><ul class="simple">
<li>ProDy has been refactored to support Python 3.  Windows installers for
Python 2.6, 2.7, 3.1, and 3.2 are available in <a class="reference internal" href="../../../manual/getprody.html#getprody"><span>Installation</span></a>.</li>
<li>Unit tests are compatible with Python 2.7 and 3.2, and running them with
other versions gives errors due to unavailability of some <a class="reference external" href="http://docs.python.org/library/unittest.html#module-unittest" title="(in Python v2.7)"><code class="xref py py-mod docutils literal"><span class="pre">unittest</span></code></a>
features.</li>
</ul>
</div></blockquote>
<p><em>Sequence Analysis</em></p>
<blockquote>
<div><ul class="simple">
<li>New applications <a class="reference internal" href="../../../manual/apps/evol/index.html#evol-apps"><span>Evol Applications</span></a> are available.</li>
<li><a class="reference internal" href="../../../manual/reference/database/pfam.html#prody.database.pfam.searchPfam" title="prody.database.pfam.searchPfam"><code class="xref py py-func docutils literal"><span class="pre">searchPfam()</span></code></a> and <a class="reference internal" href="../../../manual/reference/database/pfam.html#prody.database.pfam.fetchPfamMSA" title="prody.database.pfam.fetchPfamMSA"><code class="xref py py-func docutils literal"><span class="pre">fetchPfamMSA()</span></code></a> functions are implemented
for searching and retrieving Pfam data.  See <span class="xref std std-ref">msafiles</span> for usage
examples.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.MSAFile" title="prody.sequence.msafile.MSAFile"><code class="xref py py-class docutils literal"><span class="pre">MSAFile</span></code></a> class, <a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.parseMSA" title="prody.sequence.msafile.parseMSA"><code class="xref py py-func docutils literal"><span class="pre">parseMSA()</span></code></a> and <a class="reference internal" href="../../../manual/reference/sequence/msafile.html#prody.sequence.msafile.writeMSA" title="prody.sequence.msafile.writeMSA"><code class="xref py py-func docutils literal"><span class="pre">writeMSA()</span></code></a> functions
are implemented for reading and writing multiple sequence alignments.
See <span class="xref std std-ref">msafiles</span> for usage examples.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/msa.html#prody.sequence.msa.MSA" title="prody.sequence.msa.MSA"><code class="xref py py-class docutils literal"><span class="pre">MSA</span></code></a> class has been implemented for storing and manipulating
MSAs in memory.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.calcShannonEntropy" title="prody.sequence.analysis.calcShannonEntropy"><code class="xref py py-func docutils literal"><span class="pre">calcShannonEntropy()</span></code></a>, <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.buildMutinfoMatrix" title="prody.sequence.analysis.buildMutinfoMatrix"><code class="xref py py-func docutils literal"><span class="pre">buildMutinfoMatrix()</span></code></a>, and
<a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.calcMSAOccupancy" title="prody.sequence.analysis.calcMSAOccupancy"><code class="xref py py-func docutils literal"><span class="pre">calcMSAOccupancy()</span></code></a> functions are implemented implemented for
MSA analysis.  See <span class="xref std std-ref">msa-analysis</span> for usage examples.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/plotting.html#prody.sequence.plotting.showShannonEntropy" title="prody.sequence.plotting.showShannonEntropy"><code class="xref py py-func docutils literal"><span class="pre">showShannonEntropy()</span></code></a>, <a class="reference internal" href="../../../manual/reference/sequence/plotting.html#prody.sequence.plotting.showMutinfoMatrix" title="prody.sequence.plotting.showMutinfoMatrix"><code class="xref py py-func docutils literal"><span class="pre">showMutinfoMatrix()</span></code></a>, and
<a class="reference internal" href="../../../manual/reference/sequence/plotting.html#prody.sequence.plotting.showMSAOccupancy" title="prody.sequence.plotting.showMSAOccupancy"><code class="xref py py-func docutils literal"><span class="pre">showMSAOccupancy()</span></code></a> functions are implemented implemented for
MSA analysis.  See <span class="xref std std-ref">msa-analysis</span> for usage examples.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.applyMutinfoCorr" title="prody.sequence.analysis.applyMutinfoCorr"><code class="xref py py-func docutils literal"><span class="pre">applyMutinfoCorr()</span></code></a> and <a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.applyMutinfoNorm" title="prody.sequence.analysis.applyMutinfoNorm"><code class="xref py py-func docutils literal"><span class="pre">applyMutinfoNorm()</span></code></a> functions are
implemented for applying normalization and corrections to mutual
information matrices.</li>
<li><a class="reference internal" href="../../../manual/reference/sequence/analysis.html#prody.sequence.analysis.calcRankorder" title="prody.sequence.analysis.calcRankorder"><code class="xref py py-func docutils literal"><span class="pre">calcRankorder()</span></code></a> function is implemented for identifying highly
correlated/co-evolving pairs of residues.</li>
</ul>
</div></blockquote>
<p><strong>Bugfix</strong>:</p>
<blockquote>
<div><ul class="simple">
<li>Fixed selection issues involving use of <code class="docutils literal"><span class="pre">x</span></code> or negative numbers.</li>
</ul>
</div></blockquote>
</div>
</div>


    </div>
  </div>
</div>
<footer class="footer">
  <div class="row">
    <div class="span12">
      <p><small>
      &copy; Copyright 2010-2017, University of Pittsburgh.
      Last updated on Dec 06, 2017.
      Created using <a href="http://sphinx-doc.org">Sphinx</a> 1.3.5.
      </small></p>
    </div>
  </div>
  </div>
</footer>
<div id="codemodal" class="modal hide fade" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
  <div class="modal-header">
    <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
    <h3 id="myModalLabel">Code Snippets</h3>
  </div>
  <div class="modal-body">
    <textarea id="codesnippets"></textarea>
  </div>
  <div class="modal-footer">
    <button class="btn btn-info" id="selectcode">Select Code</button>
    <!--
    <button class="btn btn-success" id="copycode" data-clipboard-target="codesnippets">Copy Code</button>
    -->
    <button class="btn" data-dismiss="modal" aria-hidden="true">Close</button>
  </div>
  <input type="hidden" id="zeroclipboardpath" value="../../../_static/js/zeroclipboard.swf">
</div>
  </body>
</html>